### ISSUE TITLE
Support for multimedia keys

### ISSUE THREAD

**Issue by biemster** [biemster/3keys_1knob#2]
The current implementation only supports alphanumeric and KBD keys as defined from here on further: https://github.com/biemster/3keys_1knob/blob/5f5ef658ef4719743eb8c76fcf55c27fb5f310b3/include/usb_conkbd.h#L41

There are a few consumer keys that would be very handly (for example volume up/down) defined from here on:
https://github.com/biemster/3keys_1knob/blob/5f5ef658ef4719743eb8c76fcf55c27fb5f310b3/include/usb_conkbd.h#L82

Support for those would be much appreciated, but would require some rework of the internals. Especially the storage in and reading from EEPROM.

**Comment by eworm-de** [comment]
Indeed, that would be nice to have...

**Comment by eworm-de** [comment]
I had a brainstorm and hacking session with @mgisbers...

This is the official documentation: [USB HID Usage Tables - Keyboard/Keypad Page](https://usb.org/sites/default/files/hut1_4.pdf#chapter.10)
... which gives a nice table with ids to be written into `KBD_report`.

The idea on this: We ditch the ASCII to keycode mapping, as it interferes with some generic codes we want to send. After all one should be able to look up an id in a table if the same person is expected to write that id to a file with a hex editor. 😜

The eeprom layout is changed to two bytes for every key:

* first byte is a bitmap of modifiers (control, shift, alt, ...)
* second byte is the usage id found in table

Let's have some examples:

* `a` → `0x0004`
* `A` (capital `a`) → `0x0204`
* `Ctrl`+`Alt`+`Del` → `0x054c`
* Volume up → `0x0080`
* Volume down → `0x0081`

Any thoughts on this?

**Comment by eworm-de** [comment]
I've come up with a prove of concept...
https://github.com/eworm-de/3keys_1knob/tree/modifier-and-code

**Comment by biemster** [comment]
Yeah that looks nice! I was also thinking about a 2 byte layout indeed, and even a bit more elaborate since I was considering having the pad send whole sequences. So maybe the first byte can also indicate how many ids are coming?
I have to be honest though that my free time is very limited at the moment, and this project is not really high on my list, sorry.

**Comment by eworm-de** [comment]
Yes, sequences of keys (macros) would be nice... Already thinking about it.

**Comment by MrGeorgeK55** [comment]
i'm having problems with the Consumer Keyboard Keycodes, do you need a modifier in the first byte? or i'm missing something?

**Comment by biemster** [comment]
@MrGeorgeK55 instead of `KBD_{press,releaseAll}` here
https://github.com/biemster/3keys_1knob/blob/5f5ef658ef4719743eb8c76fcf55c27fb5f310b3/3keys_1knob.c#L195
and here
https://github.com/biemster/3keys_1knob/blob/5f5ef658ef4719743eb8c76fcf55c27fb5f310b3/3keys_1knob.c#L198
you need `CON_{press,releaseAll}`.

**Comment by hugokernel** [comment]
Thank you @eworm-de: Your branch works perfectly well, however, I had to make a horrible hack to be able to send the Mute keycode via the consumer functions.

I can't wait for this project to evolve a bit.

**Comment by biemster** [comment]
> I can't wait for this project to evolve a bit.

Winter is coming, and with that more indoors time so this ticket will be picked up at some point.

**Comment by hugokernel** [comment]
> * `Ctrl`+`Alt`+`Del` → `0x054c`

Could you explains how you get `0x054c` ?

I want to send this: `Ctrl` + `Win` + `l` but I don't really understand how I can do that.

Edit: Bitmap is ?|?|?|?|WIN|CTRL|SHIFT|ALT

**Comment by MisterRager** [comment]
I wanted to also add support for the consumer page keys, so I added a third configuration byte in eeprom per-key:
https://github.com/eworm-de/3keys_1knob/pull/1

**Comment by MrGeorgeK55** [comment]
> > * `Ctrl`+`Alt`+`Del` → `0x054c`
> 
> Could you explains how you get `0x054c` ?
> 
> I want to send this: `Ctrl` + `Win` + `l` but I don't really understand how I can do that.
> 
> Edit: Bitmap is ?|?|?|?|WIN|CTRL|SHIFT|ALT

So far with the results of my tests, this is the table of modifier keys:

00 None  
01 Ctrl  
02 Shift  
03 Ctrl + Shift  
04 Alt  
05 Ctrl + Alt  
06 Shift + Alt  
07 Ctrl + Shift + Alt

**Comment by MrGeorgeK55** [comment]
Also i Modified the code for a different project: [Macropad 8 Keys](https://github.com/MrGeorgeK55/Macropad-8-keys) based on CH552g, working with Keyboard keys. Consumer keys and macros up to 10 keys.  im working to adapt it to this keyboard

**Comment by MrGeorgeK55** [comment]
Came with a proof of concept of macros up to 19 consecutive keys (without any modifier by now), also with more documentation.
[Macropad-3-Keys-1-Knob](https://github.com/MrGeorgeK55/Macropad-3-keys-1-knob)

**Comment by biemster** [comment]
nice work @MrGeorgeK55! I'll gave a closer look soon when time allows

**Comment by idlelife2** [comment]
@MrGeorgeK55 @biemster 

The value of the media key playpause is 0xCD

https://github.com/rOzzy1987/MacroPad/blob/main/src/RSoft.MacroPad.BLL/Infrasturture/Model/MediaKey.cs

I found out about this link
